### PR TITLE
Fix trait bounds issue #799

### PR DESCRIPTION
**Pull request by Sympatron**
I gave #799 a try. This is what I came up with. The new tests in `defmt/tests/derive-bounds.rs` compile now.

### CODE REVIEW

**Comment by BriocheBerlin** [comment]
Thank you! We'll get back to you after the holiday break.

File: macros/src/derives/format/codegen.rs [hunk 1]
```diff
@@ -39,19 +44,16 @@ pub(crate) struct Generics<'a> {
...
-            if let GenericParam::Type(ty) = param {
-                let ident = &ty.ident;
-
-                where_clause
-                    .predicates
-                    .push(parse_quote!(#ident: defmt::Format));
-            }
+        // Extend where-clause with `Format` bounds for all field types.
+        for predicate in where_predicates {
+            where_clause.predicates.push(predicate);
```
Comment (Urhengulas) [inline macros/src/derives/format/codegen.rs:56]: You can use `.extend`: https://doc.rust-lang.org/std/iter/trait.Extend.html#tymethod.extend

File: macros/src/derives/format/codegen/enum_data.rs [hunk 2]
```diff
@@ -50,8 +53,14 @@ pub(crate) fn encode(ident: &Ident, data: &DataEnum) -> syn::Result<EncodeData>
     let stmts = vec![quote!(match self {
         #(#match_arms)*
     })];
+    // This would require syn feature "extra-traits"
+    // where_predicates.dedup_by(|a, b| a == b);
```
Comment (Urhengulas) [inline macros/src/derives/format/codegen/enum_data.rs:57]: So let's enable the feature!

**Review by Urhengulas — CHANGES_REQUESTED** [review]

File: macros/src/derives/format/codegen/fields.rs [hunk 3]
```diff
@@ -37,21 +38,30 @@ pub(crate) fn codegen(
         }
 
         let format_opt = get_defmt_format_option(field)?;
-        let ty = as_native_type(&field.ty).unwrap_or_else(|| consts::TYPE_FORMAT.to_string());
+        let ty = as_native_type(&field.ty);
+        let field_ty = if ty.is_none() { Some(&field.ty) } else { None };
+        let ty = ty.unwrap_or_else(|| consts::TYPE_FORMAT.to_string());
```
Comment (Urhengulas) [inline macros/src/derives/format/codegen/fields.rs:43]: Could you please add a comment explaining what is happening in this code block?

File: macros/src/derives/format/codegen/fields.rs [hunk 4]
```diff
@@ -37,21 +38,30 @@ pub(crate) fn codegen(
...
+            field_ty.map(|_| parse_quote!(::defmt::Format))
         } else {
             let method = format_ident!("{}", ty);
-            stmts.push(quote!(defmt::export::#method(#ident)));
+            stmts.push(quote!(::defmt::export::#method(#ident)));
+            field_ty.map(|_| parse_quote!(::defmt::Format))
+        };
+        if let Some(bound) = bound {
+            where_predicates.push(parse_quote!(#field_ty: #bound));
         }
```
Comment (Urhengulas) [inline macros/src/derives/format/codegen/fields.rs:65]: Please add a comment here as well.

File: defmt/tests/ui.rs [hunk 5]
```diff
@@ -9,5 +9,6 @@ fn ui() {
         t.compile_fail("tests/ui/*.rs");
 
         t.pass("tests/basic_usage.rs");
+        t.pass("tests/derive-bounds.rs");
```
Comment (Urhengulas) [inline defmt/tests/ui.rs:12]: Does this actually work? I don't see this test being executed with the other ui tests, but instead as a "normal" test

**Review by Urhengulas — CHANGES_REQUESTED** [review]
Looks good to me. I ran the new test through cargo expand[^1] before and after the change and the changed expansion looks sensible (see the screenshots below). Just a few nits and then we can merge imo.

baz
quux

[^1]: `cargo expand --features unstable-test --test "derive-bounds"`

**Comment by Urhengulas** [comment]
@Sympatron hey hey, do you have capacity to work on this?

File: defmt/tests/ui.rs (hunk 5 above)
Comment (Sympatron) [reply defmt/tests/ui.rs:12]: I get this output when running `cargo test --features unstable-test`:
```
[TRUNCATED]
     Running tests\ui.rs (defmt\target\debug\deps\ui-a6effd36e167384c.exe)

running 1 test
   Compiling syn v2.0.41
   Compiling thiserror-impl v1.0.50
   Compiling thiserror v1.0.50
   Compiling defmt-parser v0.3.3 (defmt\parser)
   Compiling defmt-macros v0.3.6 (defmt\macros)
   Compiling defmt v0.3.5 (defmt\defmt)
   Compiling defmt-tests v0.0.0 (defmt\target\tests\trybuild\defmt)
    Finished dev [unoptimized + debuginfo] target(s) in 9.76s


test tests\ui\derive-empty-attr.rs [should fail to compile] ... ok
test tests\ui\derive-invalid-attr-arg.rs [should fail to compile] ... ok
test tests\ui\derive-multi-attr.rs [should fail to compile] ... ok
test tests\ui\log-invalid-hint.rs [should fail to compile] ... ok
test tests\ui\panic-handler-argument.rs [should fail to compile] ... ok
test tests\ui\panic-handler-empty.rs [should fail to compile] ... ok
test tests\ui\panic-handler-export-name.rs [should fail to compile] ... ok
test tests\ui\panic-handler-no-mangle.rs [should fail to compile] ... ok
test tests\ui\write-invalid-hint.rs [should fail to compile] ... ok
test tests/basic_usage.rs [should pass] ... ok
test tests/derive-bounds.rs [should pass] ... ok


test ui ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 15.43s
[TRUNCATED]
```

So I thought this is working. Did I miss something?

File: macros/src/derives/format/codegen/fields.rs (hunk 4 above)
Comment (Sympatron) [reply macros/src/derives/format/codegen/fields.rs:68]: Done

File: macros/src/derives/format/codegen/fields.rs (hunk 3 above)
Comment (Sympatron) [reply macros/src/derives/format/codegen/fields.rs]: Done

File: macros/src/derives/format/codegen/enum_data.rs (hunk 2 above)
Comment (Sympatron) [reply macros/src/derives/format/codegen/enum_data.rs]: Done

File: macros/src/derives/format/codegen.rs (hunk 1 above)
Comment (Sympatron) [reply macros/src/derives/format/codegen.rs]: Done

**Comment by Sympatron** [comment]
@Urhengulas Sorry, somehow didn't see the notifications about your comments. Everything should be fine now.

File: macros/src/derives/format/codegen/enum_data.rs [hunk 6]
```diff
@@ -50,8 +53,14 @@ pub(crate) fn encode(ident: &Ident, data: &DataEnum) -> syn::Result<EncodeData>
     let stmts = vec![quote!(match self {
         #(#match_arms)*
     })];
+    // This would require syn feature "extra-traits"
```
Comment (Urhengulas) [inline macros/src/derives/format/codegen/enum_data.rs:56]:
```suggestion
```

File: defmt/tests/ui.rs (hunk 5 above)
Comment (Urhengulas) [reply defmt/tests/ui.rs:12]: It does work. Somehow I didn't see it

File: CHANGELOG.md [hunk 7]
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
 
+- [#800]: `defmt-macros`: Fix generic trait bounds in Format derive macro
```
Comment (Urhengulas) [inline CHANGELOG.md:10]:
```suggestion
- [#800]: `defmt-macros`: Fix generic trait bounds in Format derive macro

[#800]: https://github.com/knurling-rs/defmt/pull/800
```

File: macros/src/derives/format/codegen/fields.rs [hunk 8]
```diff
@@ -74,11 +87,11 @@ pub(crate) fn codegen(
         format_string.push(')');
     }
 
-    Ok(stmts)
+    Ok((stmts, where_predicates))
 }
 
 #[derive(Copy, Clone, Eq, PartialEq, Debug)]
-enum FormatOption {
+pub(crate) enum FormatOption {
```
Comment (Urhengulas) [inline macros/src/derives/format/codegen/fields.rs:94]:
```suggestion
enum FormatOption {
```

**Review by Urhengulas — CHANGES_REQUESTED** [review]
LGTM. I have some small changes I will take care of and then merge it.

**Review by Urhengulas — APPROVED** [review]

**Pull request merged**